plugins/notifications-backend/src/service/DefaultNotificationRecipientResolver.ts

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  Entity,
+  isGroupEntity,
+  isUserEntity,
+  parseEntityRef,
+  RELATION_HAS_MEMBER,
+  RELATION_OWNED_BY,
+  RELATION_PARENT_OF,
+  stringifyEntityRef,
+} from '@backstage/catalog-model';
+import { AuthService } from '@backstage/backend-plugin-api';
+import { CatalogService } from '@backstage/plugin-catalog-node';
+import { NotificationRecipientResolver } from '@backstage/plugin-notifications-node';
+
+const isUserEntityRef = (ref: string) =>
+  parseEntityRef(ref).kind.toLocaleLowerCase() === 'user';
+
+// Partitions array of entity references to two arrays; user entity refs and other entity refs
+const partitionEntityRefs = (refs: string[]): string[][] =>
+  refs.reduce(
+    ([userEntityRefs, otherEntityRefs]: string[][], ref: string) => {
+      return isUserEntityRef(ref)
+        ? [[...userEntityRefs, ref], otherEntityRefs]
+        : [userEntityRefs, [...otherEntityRefs, ref]];
+    },
+    [[], []],
+  );
+
+export class DefaultNotificationRecipientResolver
+  implements NotificationRecipientResolver
+{
+  constructor(
+    private readonly auth: AuthService,
+    private readonly catalog: CatalogService,
+  ) {}
+
+  async resolveNotificationRecipients(options: {
+    entityRef: string | string[] | null;
+    excludeEntityRefs?: string | string[];
```

Comment:
again, as a consumer of this interface, I'd be more helped if this was just `excludeEntityRefs: string[]`; there's no real reason to subject everyone to this variety of possible types.

And the above one could just as easily have been `entityRefs: string[]` too right? It can be an empty array if there aren't any and if that has well defined semantics.

Everybody's code will be simpler for it, right?